cdap-tms/src/main/java/co/cask/cdap/messaging/service/CoreMessagingService.java

```diff
@@ -221,22 +221,34 @@ protected void startUp() throws Exception {
     Queue<TopicId> asyncCreationTopics = new LinkedList<>();
 
     for (String topic : new HashSet<>(cConf.getTrimmedStringCollection(Constants.MessagingSystem.SYSTEM_TOPICS))) {
-      TopicId topicId;
-      try {
-        topicId = NamespaceId.SYSTEM.topic(topic);
-      } catch (IllegalArgumentException e) {
-        // Ignore invalid topic
-        LOG.warn("Ignore creation of invalid topic '{}'.", topic);
-        continue;
-      }
+      // Create topics with topic prefix and top partition
+      if (topic.contains(":")) {
+        String[] topicPrefixPartition = topic.split(":");
+        int partition;
+        if (topicPrefixPartition.length != 2) {
+          LOG.warn("Ignore creation of invalid topic '{}'. Expecting topic in the format topicPrefix:numPartitions.",
+                   topic);
+          continue;
+        }
+        String topicPrefix = topicPrefixPartition[0];
+        try {
+          partition = Integer.parseInt(topicPrefixPartition[1]);
+          if (partition <= 0) {
+            throw new NumberFormatException("metrics.messaging.partition.num must be set to a positive integer.");
+          }
+          NamespaceId.SYSTEM.topic(topicPrefix);
+        } catch (NumberFormatException e) {
+          // Ignore invalid topic
+          LOG.error(String.format(
+            "Ignore creation of invalid topic '%s' because of the invalid partition number in it.", topic), e);
+          continue;
+        }
 
-      try {
-        createTopicIfNotExists(topicId);
-      } catch (Exception e) {
-        // If failed, add it to a list so that the retry will happen asynchronously
-        LOG.warn("Topic {} creation failed with exception {}. Will retry.", topicId, e.getMessage());
-        LOG.debug("Topic {} creation failure stacktrace", topicId, e);
-        asyncCreationTopics.add(topicId);
+        for (int i = 0; i < partition; i++) {
+          createSystemTopic(topicPrefix + "_" + i, asyncCreationTopics);
```

Comment:
It's better not to add `_` implicitly. If someone needs a `_` between the prefix and the number, it is be specified in the prefix